src/coordinator/api/v1/handler/prometheus/native/read.go

```diff
@@ -114,89 +104,94 @@ func (h *PromReadHandler) read(reqCtx context.Context, w http.ResponseWriter, pa
 
 	// Results is closed by execute
 	results := make(chan executor.Query)
-	// Block series slices are sorted by start time
-	seriesMap := make(map[string][]block.Series)
 	go h.engine.ExecuteExpr(ctx, parser, opts, params, results)
 
+	// Block slices are sorted by start time
+	// TODO: Pooling
+	sortedBlockList := make([]block.Block, 0, 10)
 	for result := range results {
 		if result.Err != nil {
-			return nil, err
+			return nil, result.Err
 		}
 
 		blocks := result.Result.Blocks()
 		// TODO(nikunj): Stream blocks to client
-		for blk := range blocks {
-			iter := blk.SeriesIter()
-			for iter.Next() {
-				insertSeriesInMap(iter.Current(), seriesMap)
+		for blkResult := range blocks {
+			if blkResult.Err != nil {
+				return nil, blkResult.Err
 			}
 
-			// TODO(nikunj): Figure out how to close this outside
-			blk.Close()
+			// Insert blocks sorted by start time
+			sortedBlockList, err = insertSortedBlock(blkResult.Block, sortedBlockList)
+			if err != nil {
+				return nil, err
+			}
 		}
 	}
 
-	return seriesMapToSeriesList(seriesMap)
+	seriesList, err := sortedBlocksToSeriesList(sortedBlockList)
+	for _, b := range sortedBlockList {
+		b.Close()
+	}
+
+	return seriesList, err
 }
 
-func insertSeriesInMap(blockSeries block.Series, seriesMap map[string][]block.Series) {
-	seriesID := blockSeries.Meta.Name
-	blockList, ok := seriesMap[seriesID]
-	if !ok {
-		seriesMap[seriesID] = make([]block.Series, 1)
-		seriesMap[seriesID][0] = blockSeries
-		return
+func sortedBlocksToSeriesList(blockList []block.Block) ([]*ts.Series, error) {
+	if len(blockList) == 0 {
+		return []*ts.Series{}, nil
 	}
 
-	// Insert sorted by start time
-	for idx, s := range blockList {
-		// TODO(nikunj): Use binary search once blocklist becomes big
-		if blockSeries.Meta.Bounds.Start.Before(s.Meta.Bounds.Start) {
-			blockList = append(blockList, block.Series{})
-			copy(blockList[idx+1:], blockList[idx:])
-			blockList[idx] = blockSeries
-			seriesMap[seriesID] = blockList
-			return
-		}
+	firstBlock := blockList[0]
+	numSeries := firstBlock.Series()
+	seriesList := make([]*ts.Series, numSeries)
+	seriesIters := make([]block.SeriesIter, len(blockList))
```

Comment:
I don't think `seriesIters` is necessary; we should be able to iterate over blockList in the following loop instead

Reply:
not sure what you mean. I have added some comments to make it clear.